blog/_posts/2016-10-11-julia-0.5-highlights.md

```diff
@@ -0,0 +1,828 @@
+ -0.210096  -0.531489   0.168928  -0.128212
+ -0.13388    0.104111   0.334428   0.132699
+
+julia> mean(B,1)
+1×4 Array{Float64,2}:
+ 0.0  0.0  0.0  0.0
+{% endhighlight %}
+
+The matrix `A` is 3×4 and `mean(A,1)` is 1×4 so the `.-` operator broadcasts each column's mean along the values of the column in A, subtracting that mean from each element.
+Combining this broadcasting behavior with vectorized call syntax lets us write some fairly fancy custom array operations very concisely:
+
+{% highlight julia %}
+julia> clip.(B, [-0.3, -0.2, -0.1], [0.4, 0.3, 0.2, 0.1]')
+3×4 Array{Float64,2}:
+  0.343976   0.3       -0.3       -0.00448691
+ -0.2       -0.2        0.168928  -0.128212
+ -0.1        0.104111   0.2        0.1
+{% endhighlight %}
+
+This expression clips each element of `B` with its own specific `(hi,lo)` pair from this matrix:
+
+{% highlight julia %}
+julia> [(lo,hi) for lo=[-0.3, -0.2, -0.1], hi=[0.4, 0.3, 0.2, 0.1]]
+3×4 Array{Tuple{Float64,Float64},2}:
+ (-0.3,0.4)  (-0.3,0.3)  (-0.3,0.2)  (-0.3,0.1)
+ (-0.2,0.4)  (-0.2,0.3)  (-0.2,0.2)  (-0.2,0.1)
+ (-0.1,0.4)  (-0.1,0.3)  (-0.1,0.2)  (-0.1,0.1)
+{% endhighlight %}
+
+With the vectorized call syntax, however, the array of matrix of pairs never has to be constructed, nor does the messy code to apply `clip` to each element of `B` with the corresponding values of `lo` and `hi` need to be written.
+When `B` is larger than it is in this toy example, avoiding constructing a matrix of `(lo,hi)` pairs can be a big efficiency win.
+
+There is a bit more to the story about vectorized call syntax.
+It's common to write expressions applying multiple vectorized functions to some arrays.
+For example, one might write something like:
+
+{% highlight julia %}
+max(abs(X), abs(Y))
+{% endhighlight %}
+
+This computes the absolute values of each element of `X` and `Y` and takes the larger of the corresponding elements from `abs(X)` and `abs(Y)`.
+In this form, the code allocates two temporary intermediate arrays – one to store `abs(X)` and one for `abs(Y)`.
+If we use vectorized function call syntax, however, these calls are syntactically fused into a *single* call to `broadcast` with an anonymous function.
+In other words, `max.(abs.(X), abs.(Y))` becomes
+
+{% highlight julia %}
+broadcast((x, y)->max(abs(x), abs(y)), X, Y)
+{% endhighlight %}
+
+This version of the computation avoids allocating any intermediate arrays and performs the entire vectorized computation all at once, directly into the result array.
+We can see this difference in memory usage and speed when we benchmark these expressions:
+
+{% highlight julia %}
+julia> using BenchmarkTools
+
+julia> X, Y = rand(1000,1000), rand(1000,1000);
+
+julia> @benchmark max(abs(X), abs(Y))
+BenchmarkTools.Trial:
+  memory estimate:  22.89 mb
+  minimum time:     13.95 ms (1.77% GC)
+  median time:      14.17 ms (1.76% GC)
+  mean time:        14.32 ms (1.78% GC)
+  maximum time:     17.15 ms (3.47% GC)
+
+julia> @benchmark max.(abs.(X), abs.(Y))
```

Comment:
I don't really think this is worth getting into. Perhaps the `@benchmark` macro should do something like this automatically? Although it seems a little iffy since it makes the code that is benchmarked less like what's actually written. @johnmyleswhite?